test/set_compiler_flag.cmake

```diff
@@ -0,0 +1,66 @@
+# Copied from
```

Comment:
I find it a bit complicated for handling the `std=c++17` / `std=c++14` flags only. We try to keep the cmake as simple as possible, and eventually want to move to full modern cmake (thus not directly setting the cmake flags but rather relying on cmake API when it is possible).

What was the problem with setting `CMAKE_CXX_STANDARD` ? Although not perfect, it remains simple.

Reply:
I understand the need for simplicity, but I think this is more robust, also for setting compiler flags other than the C++ standard. 

The pattern I usually use is to have project local variables holding the flags and the flags are appended to them _after_ I've checked that they're available for the compiler at hand. In this way you can actually enforce use of certain flags at configuration rather than wait for the compiler to error out when building.
Once you have the flags in lists:
```
target_compile_options(my_target
   PUBLIC # or whatever visibility level
        "$<$<CONFIG:Debug>:${project_CXX_FLAGS_DEBUG}>"
        "$<$<CONFIG:Release>:${project_CXX_FLAGS_RELEASE}>"
   )
```
Another related point is that CMake doesn't offer every imaginable compiler flag as a property to be set on targets. As far as I know, PIC, IPO, C/C++ standard, and symbol visibility can be set that way.

Regarding the C++ standard. `set(CMAKE_CXX_STANDARD 17)` will only work with CMake >= 3.8 With earlier versions, I am not sure what happens. It will probably be set to C++14. In this particular case it means we still need to check whether `-std=c++17` is available by hand. I don't see much point in relying on `CMAKE_CXX_STANDARD` for C++14 then.

Reply:
Alright, makes sense.